File: python/src/bsb/client_generator.py
from __future__ import annotations

import json
import keyword
import re
import tomllib
from pathlib import Path
from typing import Any

from .schema_events import FLIP_MAP, import_event_schemas


def event_name_to_method_name(name: str) -> str:
    value = re.sub(r"[^0-9a-zA-Z]+", "_", name).strip("_").lower() or "event"
    if value[0].isdigit():
        value = "event_" + value
    return value + "_" if keyword.iskeyword(value) else value


def plugin_name_to_class_name(name: str) -> str:
    parts = re.findall(r"[A-Za-z0-9]+", name.removeprefix("service-"))
    value = "".join(part[:1].upper() + part[1:] for part in parts) or "Plugin"
    return ("Plugin" if value[0].isdigit() else "") + value + "Client"


def generate_client_code(schema_export: dict[str, Any], plugin_id: str) -> str:
    import_event_schemas(schema_export)
    class_name = plugin_name_to_class_name(plugin_id)
    declarations: list[str] = []
    methods: list[str] = []
    used = {"__init__", "events", "context", "target_plugin_name", "schema"}
    types = {class_name}

    def shape(document: dict, hint: str) -> str:
        definitions = document.get("definitions", {})
        references = {key: f"{hint}Definition{i}" for i, key in enumerate(definitions)}
        seen = set()

        def native(node: dict, name: str) -> str:
            kind = node["kind"]
            if kind == "string":
                return "str"
            if kind == "bool":
                return "bool"
            if kind in ("int", "int8", "uint8", "int16", "uint16", "int32", "uint32", "int64", "uint64"):
                return "int"
            if kind in ("number", "float32", "float64"):
                return "float"
            if kind in ("any", "unknown"):
                return "Any"
            if kind == "null":
                return "None"
            if kind == "never":
                return "Never"
            if kind in ("optional", "nullable"):
                return f"Union[{native(node.get('schema', node.get('inner')), name)}, None]"
            if kind == "array":
                return f"list[{native(node['items'], name + 'Item')}]"
            if kind == "record":
                return f"dict[str, {native(node.get('valueSchema', node.get('values')), name + 'Value')}]"
            if kind == "tuple":
                elements = [native(element, name + str(i)) for i, element in enumerate(node.get("elements", node.get("items", [])))]
                return "tuple[" + (", ".join(elements) if elements else "()") + "]"
            if kind == "union":
                return "Union[" + ", ".join(native(value, name + "Variant" + str(i)) for i, value in enumerate(node.get("variants", node.get("schemas", [])))) + "]"
            if kind == "intersection":
                values = node.get("allOf", node.get("schemas", []))
                if all(value["kind"] == "object" for value in values):
                    properties = {key: value for member in values for key, value in member.get("properties", {}).items()}
                    required = list({key for member in values for key in member.get("required", [])})
                    return native({"kind": "object", "properties": properties, "required": required}, name)
                return native(values[-1], name)
            if kind in ("enum", "literal"):
                values = node["values"] if kind == "enum" else [node.get("value")]
                parts = ["float" if type(value) is float else f"Literal[{value!r}]" for value in values]
                return parts[0] if len(parts) == 1 else "Union[" + ", ".join(parts) + "]"
            if kind == "ref":
                key = node["ref"].removeprefix("#/definitions/")
                if key not in references:
                    raise ValueError(f"Unknown schema definition {key}")
                reference = references[key]
                if key not in seen:
                    seen.add(key)
                    resolved = native(definitions[key], reference)
                    if resolved != reference:
                        declarations.append(f"{reference}: TypeAlias = {resolved}")
                return repr(reference)
            if kind == "object":
                if name in types:
                    raise ValueError(f"Generated type collision {name}")
                types.add(name)
                required = set(node.get("required", []))
                fields = {}
                for key, value in node.get("properties", {}).items():
                    optional = key not in required or value["kind"] == "optional"
                    item_type = native(value.get("schema", value.get("inner")) if value["kind"] == "optional" else value, name + plugin_name_to_class_name(key).removesuffix("Client"))
                    fields[key] = f"NotRequired[{item_type}]" if optional else item_type
                declarations.append(f"{name} = TypedDict({name!r}, {fields!r})")
                return name
            raise ValueError(f"Unsupported client schema kind {kind}")

        return native(document["root"], hint)

    for event, definition in sorted(schema_export["events"].items()):
        category = FLIP_MAP[definition["category"]]
        method = event_name_to_method_name(event)
        listens = category.startswith("on")
        if listens:
            method = "on_" + method
        elif category == "emitBroadcast":
            method = "emit_" + method
        if method in used:
            raise ValueError(f"Generated method collision {method}")
        used.add(method)
        hint = class_name + plugin_name_to_class_name(method).removesuffix("Client")
        input_type = shape(definition["inputSchema"], hint + "Input")
        output_type = shape(definition["outputSchema"], hint + "Output") if definition.get("outputSchema") is not None else "None"
        specific = "" if definition["type"] == "broadcast" else ", server_id: str | None = None"
        forward = "" if definition["type"] == "broadcast" else ", server_id=server_id"
        if listens:
            operation = {"onEvents": "on_event", "onReturnableEvents": "on_returnable_event", "onBroadcast": "on_broadcast"}[category]
            methods.extend([f"    async def {method}(self, handler: Callable[[Trace, {input_type}], Awaitable[{output_type}]], *, obs: Trace | None = None{specific}) -> None:",
                "        async def wrapped(trace, payload):",
                f"            return await handler(trace, client_value(self.schema['events'][{event!r}]['inputSchema'], payload))",
                f"        await self.events.{operation}({event!r}, wrapped, obs=obs{forward})", ""])
        elif category == "emitReturnableEvents":
            timeout = definition.get("defaultTimeout", 5)
            methods.extend([f"    async def {method}(self, payload: {input_type}, timeout_seconds: float = {timeout!r}, *, obs: Trace | None = None{specific}) -> {output_type}:",
                f"        value = await self.events.emit_event_and_return({event!r}, payload, timeout_seconds=timeout_seconds, obs=obs{forward})",
                f"        return cast({output_type!r}, client_value(self.schema['events'][{event!r}]['outputSchema'], value))", ""])
        else:
            operation = "emit_event" if category == "emitEvents" else "emit_broadcast"
            methods.extend([f"    async def {method}(self, payload: {input_type}, *, obs: Trace | None = None{specific}) -> None:",
                f"        await self.events.{operation}({event!r}, payload, obs=obs{forward})", ""])
    target = schema_export.get("pluginId") or schema_export.get("pluginName") or plugin_id
    return "\n".join([
        "# Generated by BSB from the saved portable schema.", "from __future__ import annotations", "import json",
        "from typing import Any, Awaitable, Callable, Literal, Never, NotRequired, TypeAlias, TypedDict, Union, cast",
        "from bsb.base import BSBService, ServiceClient", "from bsb.observable import Trace", "from bsb.schema import client_value", "",
        *declarations, "", f"class {class_name}(ServiceClient):",
        f"    schema = json.loads({json.dumps(schema_export, ensure_ascii=True)!r})", "",
        "    def __init__(self, context: BSBService, target_plugin_name: str | None = None) -> None:",
        f"        super().__init__(target_plugin_name or {target!r}, context, self.schema)", "", *methods])


def ensure_generated_layout(project_root: str | Path) -> tuple[Path, Path]:
    root = Path(project_root)
    schemas = root / ".bsb" / "schemas"
    project = tomllib.loads((root / "pyproject.toml").read_text(encoding="utf-8")) if (root / "pyproject.toml").exists() else {}
    package = project.get("tool", {}).get("bsb", {}).get("clients-package", "bsb_clients")
    if not isinstance(package, str) or any(not part.isidentifier() or keyword.iskeyword(part) for part in package.split(".")):
        raise ValueError("Invalid generated clients package")
    clients = (root / "src" if (root / "src").is_dir() else root).joinpath(*package.split("."))
    schemas.mkdir(parents=True, exist_ok=True)
    clients.mkdir(parents=True, exist_ok=True)
    return schemas, clients


def validate_client_names(values) -> None:
    names, classes = set(), set()
    for value in values:
        name, cls = event_name_to_method_name(value), plugin_name_to_class_name(value)
        if name in names or cls in classes:
            raise ValueError("Installed client names collide after Python normalization")
        names.add(name)
        classes.add(cls)


def generate_clients(project_root: str | Path) -> list[Path]:
    schemas_dir, clients_dir = ensure_generated_layout(project_root)
    files = sorted(schemas_dir.glob("*.json"))
    legacy = Path(project_root) / "src" / ".bsb" / "schemas"
    files += sorted(legacy.glob("*.json"))
    validate_client_names([file.stem for file in files])
    generated = []
    for file in files:
        name = event_name_to_method_name(file.stem)
        cls = plugin_name_to_class_name(file.stem)
        code = generate_client_code(json.loads(file.read_text(encoding="utf-8")), file.stem)
        generated.append((clients_dir / (name + ".py"), cls, code))
    current = {path for path, _, _ in generated}
    for path in clients_dir.glob("*.py"):
        if path.name != "__init__.py" and path not in current and path.read_text(encoding="utf-8").startswith("# Generated by BSB from the saved portable schema.\n"):
            path.unlink()
    for path, _, code in generated:
        path.write_text(code, encoding="utf-8")
    (clients_dir / "__init__.py").write_text("\n".join(f"from .{path.stem} import {cls} as {cls}" for path, cls, _ in generated) + "\n", encoding="utf-8")
    return [path for path, _, _ in generated]

File: dotnet/BSB/Tooling/ClientGenerator.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using BSB.Interfaces;

namespace BSB.Tooling;

/// <summary>Produces native clients for any registry service, independently of its implementation language.</summary>
public static class ClientGenerator
{
    public static string Identifier(string name)
    {
        var words = Regex.Split(name, "[^A-Za-z0-9]+").Where(x => x.Length > 0);
        var result = string.Concat(words.Select(x => char.ToUpperInvariant(x[0]) + x[1..]));
        return result.Length == 0 ? "Generated" : char.IsDigit(result[0]) ? "_" + result : result;
    }
    private static string Quote(string value) => JsonSerializer.Serialize(value);

    public static string Generate(EventSchemaExport export, string localName)
    {
        BSBEventSchemas.Import(export); // Reject invalid schemas before producing source.
        var name = Identifier(localName) + "Client";
        var declarations = new StringBuilder();
        var methods = new StringBuilder();
        var used = new HashSet<string>(StringComparer.Ordinal) { name, "Events", "Schema", "Convert" };
        var typeNames = new HashSet<string>(StringComparer.Ordinal) { name };

        string Shape(JsonObject document, string hint)
        {
            var definitions = document["definitions"]?.AsObject() ?? new();
            var references = definitions.Select((p, index) => (p.Key, Name: hint + "Definition" + index)).ToDictionary(p => p.Key, p => p.Name);
            var emitted = new HashSet<string>();
            string Type(JsonNode node, string suggested)
            {
                var kind = node["kind"]?.GetValue<string>();
                switch (kind)
                {
                    case "string": return "string";
                    case "bool": return "bool";
                    case "int8": return "sbyte";
                    case "uint8": return "byte";
                    case "int16": return "short";
                    case "uint16": return "ushort";
                    case "int32": return "int";
                    case "uint32": return "uint";
                    case "int64": return "long";
                    case "uint64": return "ulong";
                    case "float32": return "float";
                    case "int": case "number": case "float64": return "double";
                    case "null": case "any": case "unknown": case "never": return "JsonElement";
                    case "literal": return node["value"]?.GetValueKind() switch { JsonValueKind.String => "string", JsonValueKind.True or JsonValueKind.False => "bool", JsonValueKind.Number => "double", _ => "JsonElement" };
                    case "nullable": case "optional":
                        var inner = Type(node["inner"]!, suggested);
                        return inner.EndsWith('?') ? inner : inner + "?";
                    case "array": return $"List<{Type(node["items"]!, suggested + "Item")}>";
                    case "record": return $"Dictionary<string, {Type(node["valueSchema"] ?? node["values"]!, suggested + "Value")}>";
                    case "ref":
                        var referenceKey = node["ref"]!.GetValue<string>();
                        if (!references.TryGetValue(referenceKey, out var reference)) throw new JsonException($"Unknown schema definition: {referenceKey}");
                        if (emitted.Add(referenceKey))
                        {
                            references[referenceKey] = Type(definitions[referenceKey]!, reference);
                        }
                        else if (references[referenceKey] == reference && !typeNames.Contains(reference))
                            throw new JsonException($"Recursive schema alias must resolve through a named object: {referenceKey}");
                        return references[referenceKey];
                    case "enum":
                        if (node["values"]!.AsArray().Any(x => x?.GetValueKind() != JsonValueKind.String)) return "JsonElement";
                        if (!typeNames.Add(suggested)) throw new JsonException($"Generated type collision: {suggested}");
                        declarations.AppendLine($"[JsonConverter(typeof(JsonStringEnumConverter<{suggested}>))]\npublic enum {suggested}\n{{");
                        var values = new HashSet<string>();
                        foreach (var value in node["values"]!.AsArray())
                        {
                            var text = value!.GetValue<string>(); var member = Identifier(text);
                            if (!values.Add(member)) throw new JsonException($"Generated enum collision: {text}");
                            declarations.AppendLine($"    [JsonStringEnumMemberName({Quote(text)})] {member},");
                        }
                        declarations.AppendLine("}");
                        return suggested;
                    case "object":
                        if (!typeNames.Add(suggested)) throw new JsonException($"Generated type collision: {suggested}");
                        var properties = new StringBuilder();
                        var propertyNames = new HashSet<string> { suggested };
                        var required = node["required"]?.AsArray().Select(x => x!.GetValue<string>()).ToHashSet() ?? new();
                        foreach (var (key, value) in node["properties"]!.AsObject())
                        {
                            var property = Identifier(key);
                            if (!propertyNames.Add(property)) throw new JsonException($"Generated property collision: {key}");
                            var optional = !required.Contains(key) || value!["kind"]!.GetValue<string>() == "optional";
                            var propertyNode = value!;
                            while (propertyNode["kind"]?.GetValue<string>() == "optional") propertyNode = propertyNode["inner"]!;
                            var propertyType = Type(propertyNode, suggested + property);
                            if (optional) propertyType = $"OptionalValue<{propertyType}>";
                            properties.AppendLine($"    [JsonPropertyName({Quote(key)})]");
                            if (optional) properties.AppendLine("    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]");
                            properties.AppendLine($"    public {(optional ? "" : "required ")}{propertyType} {property} {{ get; init; }}");
                        }
                        if (node["unknownKeys"]?.GetValue<string>() == "passthrough")
                            properties.AppendLine("    [JsonExtensionData] public Dictionary<string, JsonElement>? AdditionalProperties { get; init; }");
                        declarations.AppendLine($"public sealed record {suggested}\n{{\n{properties}}}");
                        return suggested;
                    case "union": case "intersection": case "tuple":
                        // C# has no general structural union/intersection type. The complete schema still validates this JSON value.
                        return "JsonElement";
                    default: throw new JsonException($"Unsupported schema kind: {kind}");
                }
            }
            return Type(document["root"]!, hint);
        }

        foreach (var (eventName, definition) in export.Events)
        {
            var method = Identifier(eventName);
            var listens = definition.Category.StartsWith("emit", StringComparison.Ordinal);
            if (listens) method = "On" + method;
            if (!used.Add(method)) throw new JsonException($"Generated method collision: {eventName}");
            var input = Shape(definition.InputSchema!, name + method + "Input");
            var output = definition.OutputSchema is null ? "object?" : Shape(definition.OutputSchema, name + method + "Output");
            var key = Quote(eventName);
            var timeout = definition.DefaultTimeoutSeconds ?? 5;
            if (listens)
            {
                var operation = definition.Type switch { "returnable" => "OnReturnableEvent", "broadcast" => "OnBroadcast", _ => "OnEvent" };
                var result = definition.Type == "returnable" ? $"Task<{output}>" : "Task";
                var lambda = definition.Type == "returnable" ? $"async (trace, value) => (object?)await handler(trace, Convert<{input}>(value))" : $"(trace, value) => handler(trace, Convert<{input}>(value))";
                methods.AppendLine($"    public Task {method}(IObservable obs, Func<IObservable, {input}, {result}> handler) => _events.{operation}({key}, obs, {lambda});");
                if (definition.Type != "broadcast")
                {
                    if (!used.Add(method + "Specific")) throw new JsonException($"Generated method collision: {eventName}Specific");
                    methods.AppendLine($"    public Task {method}Specific(string serverId, IObservable obs, Func<IObservable, {input}, {result}> handler) => _events.{operation}Specific(serverId, {key}, obs, {lambda});");
                }
            }
            else if (definition.Type == "returnable")
            {
                methods.AppendLine($"    public async Task<{output}> {method}(IObservable obs, {input} input, int timeoutSeconds = {timeout}) => Convert<{output}>(await _events.EmitEventAndReturn({key}, obs, input, timeoutSeconds));");
                if (!used.Add(method + "Specific")) throw new JsonException($"Generated method collision: {eventName}Specific");
                methods.AppendLine($"    public async Task<{output}> {method}Specific(string serverId, IObservable obs, {input} input, int timeoutSeconds = {timeout}) => Convert<{output}>(await _events.EmitEventAndReturnSpecific(serverId, {key}, obs, input, timeoutSeconds));");
            }
            else
            {
                methods.AppendLine($"    public Task {method}(IObservable obs, {input} input) => _events.{(definition.Type == "broadcast" ? "EmitBroadcast" : "EmitEvent")}({key}, obs, input);");
                if (definition.Type != "broadcast")
                {
                    if (!used.Add(method + "Specific")) throw new JsonException($"Generated method collision: {eventName}Specific");
                    methods.AppendLine($"    public Task {method}Specific(string serverId, IObservable obs, {input} input) => _events.EmitEventSpecific(serverId, {key}, obs, input);");
                }
            }
        }
        return $$"""
        // Generated by BSB. Regenerate from the saved registry schema.
        #nullable enable
        using System;
        using System.Collections.Generic;
        using System.Threading.Tasks;
        using System.Text.Json;
        using System.Text.Json.Serialization;
        using BSB.Base;
        using BSB.Interfaces;

        {{declarations}}
        public sealed class {{name}}
        {
            private readonly PluginEvents _events;
            private static readonly EventSchemaExport Schema = EventSchemaExport.FromJson({{Quote(export.ToJson())}});
            public {{name}}(PluginEvents context, string targetPlugin = {{Quote(export.PluginId ?? export.PluginName)}})
                => _events = context.CreateClient(targetPlugin, Schema);
            private static T Convert<T>(object? value) => JsonSerializer.Deserialize<T>(JsonSerializer.Serialize(value), EventSchemaExport.JsonOptions)!;
        {{methods}}
        }
        """;
    }
}
